File: publishable-key-not-linked-to-a-sales-channel/python/link_publishable_key.py
"""Find Medusa publishable keys with zero active sales-channel links and link them, safely.

A publishable key only scopes /store/* requests through a link to one or more sales
channels. If that link is missing, or every linked channel is disabled, the key
resolves to zero sales channels and the storefront sees no products. This lists
every publishable key, decides what to do with a pure function, and only writes
when a target sales channel is explicit or there is exactly one unambiguous
default channel. Every other case is reported only. Run once, or on a schedule.
Safe to run again and again.

Guide: https://www.allanninal.dev/medusa/publishable-key-not-linked-to-a-sales-channel/
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("link_publishable_key")

BACKEND_URL = os.environ["MEDUSA_BACKEND_URL"]
ADMIN_EMAIL = os.environ["MEDUSA_ADMIN_EMAIL"]
ADMIN_PASSWORD = os.environ["MEDUSA_ADMIN_PASSWORD"]
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
SALES_CHANNEL_ID_OVERRIDE = os.environ.get("SALES_CHANNEL_ID", "").strip() or None


def get_admin_token():
    r = requests.post(
        f"{BACKEND_URL}/auth/user/emailpass",
        json={"email": ADMIN_EMAIL, "password": ADMIN_PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def get_publishable_keys(token):
    r = requests.get(
        f"{BACKEND_URL}/admin/api-keys",
        headers={"Authorization": f"Bearer {token}"},
        params={"type": "publishable", "fields": "id,token,title,revoked_at,*sales_channels"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["api_keys"]


def get_enabled_sales_channels(token):
    r = requests.get(
        f"{BACKEND_URL}/admin/sales-channels",
        headers={"Authorization": f"Bearer {token}"},
        params={"fields": "id,name,is_disabled"},
        timeout=30,
    )
    r.raise_for_status()
    channels = r.json()["sales_channels"]
    return [c for c in channels if not c.get("is_disabled")]


def decide_api_key_repair(api_key, default_sales_channel_id):
    """Pure decision function. No I/O.

    api_key: {"id": str, "revoked_at": str | None, "sales_channels": [{"id": str, "is_disabled": bool}, ...]}
    default_sales_channel_id: str | None

    Returns {"action": "none" | "flag" | "link", "reason": str, "sales_channel_id_to_add"?: str}.
    """
    if api_key.get("revoked_at"):
        return {"action": "none", "reason": "key revoked"}

    active_links = [sc for sc in api_key.get("sales_channels") or [] if not sc.get("is_disabled")]
    if len(active_links) > 0:
        return {"action": "none", "reason": "already linked to an active sales channel"}

    if default_sales_channel_id is None:
        return {"action": "flag", "reason": "no sales channel linked and no unambiguous default to link"}

    return {
        "action": "link",
        "reason": "key has zero active sales-channel links",
        "sales_channel_id_to_add": default_sales_channel_id,
    }


def link_sales_channel(token, api_key_id, sales_channel_id):
    r = requests.post(
        f"{BACKEND_URL}/admin/api-keys/{api_key_id}/sales-channels",
        headers={"Authorization": f"Bearer {token}", "Content-Type": "application/json"},
        json={"add": [sales_channel_id]},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def confirm_linked(token, api_key_id, sales_channel_id):
    r = requests.get(
        f"{BACKEND_URL}/admin/api-keys/{api_key_id}",
        headers={"Authorization": f"Bearer {token}"},
        params={"fields": "id,*sales_channels"},
        timeout=30,
    )
    r.raise_for_status()
    linked_ids = {sc["id"] for sc in r.json()["api_key"]["sales_channels"]}
    return sales_channel_id in linked_ids


def default_sales_channel_id(token):
    if SALES_CHANNEL_ID_OVERRIDE:
        return SALES_CHANNEL_ID_OVERRIDE
    enabled = get_enabled_sales_channels(token)
    if len(enabled) == 1:
        return enabled[0]["id"]
    return None


def run():
    token = get_admin_token()
    keys = get_publishable_keys(token)
    default_sc_id = default_sales_channel_id(token)

    linked = 0
    flagged = 0
    for api_key in keys:
        decision = decide_api_key_repair(api_key, default_sc_id)
        log.info("Key %s (%s): action=%s reason=%s", api_key["id"], api_key.get("title"), decision["action"], decision["reason"])

        if decision["action"] == "flag":
            flagged += 1
            continue
        if decision["action"] != "link":
            continue

        sc_id = decision["sales_channel_id_to_add"]
        log.info(
            "%s api key %s to sales channel %s",
            "Would link" if DRY_RUN else "Linking", api_key["id"], sc_id,
        )
        if not DRY_RUN:
            link_sales_channel(token, api_key["id"], sc_id)
            if not confirm_linked(token, api_key["id"], sc_id):
                raise RuntimeError(f"Link did not take effect for key {api_key['id']}")
            log.info("Confirmed. Key %s is now linked to sales channel %s.", api_key["id"], sc_id)
        linked += 1

    log.info("Done. %d key(s) %s, %d key(s) flagged for review.", linked, "to link" if DRY_RUN else "linked", flagged)


if __name__ == "__main__":
    run()

File: publishable-key-not-linked-to-a-sales-channel/node/link-publishable-key.js
/**
 * Find Medusa publishable keys with zero active sales-channel links and link them, safely.
 *
 * A publishable key only scopes /store/* requests through a link to one or more sales
 * channels. If that link is missing, or every linked channel is disabled, the key
 * resolves to zero sales channels and the storefront sees no products. This lists
 * every publishable key, decides what to do with a pure function, and only writes
 * when a target sales channel is explicit or there is exactly one unambiguous
 * default channel. Every other case is reported only. Run once, or on a schedule.
 *
 * Guide: https://www.allanninal.dev/medusa/publishable-key-not-linked-to-a-sales-channel/
 */
import { pathToFileURL } from "node:url";

const BACKEND_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const ADMIN_EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const ADMIN_PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";
const SALES_CHANNEL_ID_OVERRIDE = (process.env.SALES_CHANNEL_ID || "").trim() || null;

export function decideApiKeyRepair(apiKey, defaultSalesChannelId) {
  if (apiKey.revoked_at) {
    return { action: "none", reason: "key revoked" };
  }

  const activeLinks = (apiKey.sales_channels || []).filter((sc) => !sc.is_disabled);
  if (activeLinks.length > 0) {
    return { action: "none", reason: "already linked to an active sales channel" };
  }

  if (defaultSalesChannelId === null || defaultSalesChannelId === undefined) {
    return { action: "flag", reason: "no sales channel linked and no unambiguous default to link" };
  }

  return {
    action: "link",
    reason: "key has zero active sales-channel links",
    salesChannelIdToAdd: defaultSalesChannelId,
  };
}

async function getSdk() {
  const { default: Medusa } = await import("@medusajs/js-sdk");
  const sdk = new Medusa({ baseUrl: BACKEND_URL, auth: { type: "jwt" } });
  await sdk.auth.login("user", "emailpass", { email: ADMIN_EMAIL, password: ADMIN_PASSWORD });
  return sdk;
}

async function getPublishableKeys(sdk) {
  const { api_keys } = await sdk.admin.apiKey.list({
    type: "publishable",
    fields: "id,token,title,revoked_at,*sales_channels",
  });
  return api_keys;
}

async function getEnabledSalesChannels(sdk) {
  const { sales_channels } = await sdk.admin.salesChannel.list({
    fields: "id,name,is_disabled",
  });
  return sales_channels.filter((c) => !c.is_disabled);
}

async function defaultSalesChannelId(sdk) {
  if (SALES_CHANNEL_ID_OVERRIDE) return SALES_CHANNEL_ID_OVERRIDE;
  const enabled = await getEnabledSalesChannels(sdk);
  return enabled.length === 1 ? enabled[0].id : null;
}

async function linkSalesChannel(sdk, apiKeyId, salesChannelId) {
  return sdk.admin.apiKey.batchSalesChannels(apiKeyId, { add: [salesChannelId] });
}

async function confirmLinked(sdk, apiKeyId, salesChannelId) {
  const { api_key } = await sdk.admin.apiKey.retrieve(apiKeyId, { fields: "id,*sales_channels" });
  const linkedIds = new Set(api_key.sales_channels.map((sc) => sc.id));
  return linkedIds.has(salesChannelId);
}

export async function run() {
  const sdk = await getSdk();
  const keys = await getPublishableKeys(sdk);
  const defaultScId = await defaultSalesChannelId(sdk);

  let linked = 0;
  let flagged = 0;
  for (const apiKey of keys) {
    const decision = decideApiKeyRepair(apiKey, defaultScId);
    console.log(`Key ${apiKey.id} (${apiKey.title}): action=${decision.action} reason=${decision.reason}`);

    if (decision.action === "flag") {
      flagged++;
      continue;
    }
    if (decision.action !== "link") continue;

    const scId = decision.salesChannelIdToAdd;
    console.log(`${DRY_RUN ? "Would link" : "Linking"} api key ${apiKey.id} to sales channel ${scId}`);
    if (!DRY_RUN) {
      await linkSalesChannel(sdk, apiKey.id, scId);
      const ok = await confirmLinked(sdk, apiKey.id, scId);
      if (!ok) throw new Error(`Link did not take effect for key ${apiKey.id}`);
      console.log(`Confirmed. Key ${apiKey.id} is now linked to sales channel ${scId}.`);
    }
    linked++;
  }

  console.log(`Done. ${linked} key(s) ${DRY_RUN ? "to link" : "linked"}, ${flagged} key(s) flagged for review.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
